replica/table.cc

```diff
@@ -861,9 +863,19 @@ table::seal_active_memtable(compaction_group& cg, flush_permit&& flush_permit) n
                 ex = std::current_exception();
                 _config.cf_stats->failed_memtables_flushes_count++;
 
-                auto should_retry = [](auto* ep) {
-                    int ec = ep->code().value();
-                    return ec == ENOSPC || ec == EDQUOT;
+                auto should_retry = overloaded_functor{
+                    [](std::system_error* ep) {
+                        int ec = ep->code().value();
+                        return ec == ENOSPC || ec == EDQUOT;
+                    },
+                    [](storage_io_error* ep) {
+                        int ec = ep->code().value();
+                        return ec == ENOSPC || ec == EDQUOT;
+                    },
+                    [] (seastar::httpd::unexpected_status_error* ep) {
+                        auto sc = ep->status();
+                        return sc == seastar::http::reply::status_type::forbidden;
+                    }
```

Comment:
what about internal_server_error?

https://docs.aws.amazon.com/AmazonS3/latest/userguide/ErrorBestPractices.html

you probably thought of segregating error handling per storage type, but it may not be worth the effort. may bring obscurity.

Reply:
The internal server error is to be handled at another level. E.g. when scylla tries to read TOC file and hits 500 code it should retry it as well. This change is about retrying memtable flush that happens due to misconfiguration, because that's a type of errors that happens long after you configured and started scylla and that doesn't heal itself eventually unless someone does something _locally_

Reply:
> you probably thought of segregating error handling per storage type, but it may not be worth the effort. may bring obscurity.

Yes, absolutely. My first though was to introduce smth like `retriable_flush_error` exception, but I realized that it's too deep refactoring that will anyway end up with top-level catching of ENOSPC/EDQUOT/http.forbidden. So indeed not worth it

Reply:
This looks like a layering violation. What does memtable know about http errors here? It should be caught be the storage layer and translated into some other error (misconfigured_storage_error or something).

Exceptions are a bad way to communicate discriminatable errors since they don't appear in any signature, but that's life.

Reply:
> What does memtable know about http errors here?

Memtable knows that the lack of space on the disk is expressed by system_error(ENOSPC/EDQUOT)
Memtable knows that misconfigured S3 storage is expressed by unexpected_status_error(forbidden)

Why is either of them is better/worse than its counterpart?

> It should be caught be the storage layer and translated into some other error (misconfigured_storage_error or something).

But that's what this place does -- it converts lower-level exceptions into higher-level ~~exceptions~~ decision to retry. Similarly we can ~~sweep this problem under the rug~~ resolve this layering violation by moving the `with_retry` lambda somewhere into util/ or sstables/ and name it respectively. I feel there's something wrong with this rename, but what?

Reply:
system_error - ENOSPC is standard and means we ran out of space. It would make sense to convert S3 out of space to this error, if s3 could run out of space.

unexpected_status_error(forbidden) means what? we can imagine that in one http transport it's retryable and in another it's fatal.

This place is memtable, not something that knows about individual transports.

I'm willing to overlook it since it's unlikely to cause trouble, but I do think it's a layering violation.

Reply:
> unexpected_status_error(forbidden) means what? 

It means some policy forbids flushing a memtable. I wonder what it we `chmod -r 0000 /var/lib/scylla` after it starts but before it tries to flush a memtable. It could've been system_error(EPERM/EACCESS) check next to ENOSPC/EDQUOT one to also wait for system administrator putting (putting _back_, since distributed loader performs `verify_owner_and_mode` in advance) proper permissions on /var/lib/scylla.